include/deal.II/differentiation/ad/ad_drivers.h

```diff
@@ -0,0 +1,1124 @@
+      const unsigned int &,
+      const unsigned int &,
+      const std::vector<ScalarType> &,
+      Vector<ScalarType> &)
+    {
+      AssertThrow(false, ExcNotImplemented());
+    }
+
+
+    template<typename ADNumberType, typename ScalarType, typename T>
+    void
+    TapedDrivers<ADNumberType,ScalarType,T>::hessian (
+      const unsigned int &,
+      const unsigned int &,
+      const std::vector<ScalarType> &,
+      FullMatrix<ScalarType> &)
+    {
+      AssertThrow(false, ExcNotImplemented());
+    }
+
+
+    template<typename ADNumberType, typename ScalarType, typename T>
+    void
+    TapedDrivers<ADNumberType,ScalarType,T>::values (
+      const unsigned int &,
+      const unsigned int &,
+      const unsigned int &,
+      const std::vector<ScalarType> &,
+      Vector<ScalarType> &)
+    {
+      AssertThrow(false, ExcNotImplemented());
+    }
+
+
+    template<typename ADNumberType, typename ScalarType, typename T>
+    void
+    TapedDrivers<ADNumberType,ScalarType,T>::jacobian (
+      const unsigned int &,
+      const unsigned int &,
+      const unsigned int &,
+      const std::vector<ScalarType> &,
+      FullMatrix<ScalarType> &)
+    {
+      AssertThrow(false, ExcNotImplemented());
+    }
+
+
+    // Specialization for taped Adol-C auto-differentiable numbers.
+    //
+    // Note: In the case of Adol-C taped numbers, the associated scalar
+    // type is always expected to be a double. So we need to make a further
+    // specialization when ScalarType is a float.
+    template<typename ADNumberType>
+    struct TapedDrivers<ADNumberType,double,typename std::enable_if<
+      ADNumberTraits<ADNumberType>::type_code == NumberTypes::adolc_taped
+      >::type>
+    {
+      typedef double scalar_type;
+
+      // === Taping ===
+
+      static void
+      enable_taping(const unsigned int &tape_index,
+                    const bool         &keep)
+      {
+        trace_on(tape_index,keep);
+      }
+
+      static void
+      disable_taping(const unsigned int &active_tape_index,
+                     const bool         &write_tapes_to_file)
+      {
+        if (write_tapes_to_file)
+          {
+            trace_off(active_tape_index); // Slow
+
+            std::vector<std::size_t> counts (STAT_SIZE);
+#ifdef DEAL_II_WITH_ADOLC
+            ::tapestats(active_tape_index, counts.data());
+#else
+            AssertThrow(false, ExcNotImplemented());
+#endif
+          }
+        else
+          trace_off(); // Fast(er)
+      }
+
+      // === Scalar drivers ===
+
+      static scalar_type
+      value (const unsigned int             &active_tape_index,
+             const unsigned int             &n_independent_variables,
+             const std::vector<scalar_type> &independent_variables)
+      {
+        Assert(independent_variables.size() == n_independent_variables,
+               ExcDimensionMismatch(independent_variables.size(),n_independent_variables));
+
+        double *f = new double();
```

Comment:
why not `double values = 0.;` and then give address to it for adolc? Then you don't need cleanup nor temporary variable. Although here you assume that `double` is castable to `scalar_type`. What if scalar_type is `std::complex<double>`?

Reply:
Good idea. My implementation was aligned with what's suggested in their documentation, but you're right that this can be optimised in a few places. I've done so here and where possible below.

Reply:
BTW. `scalar_type` is always double for this class (see typedef above). I've changed the remaining instances of `double` to `scalar_type` though to be consistent. Thanks for pointing this out!